src/commons/crypto/signers/kmip/internal.rs

```diff
@@ -0,0 +1,776 @@
+            return Err(SignerError::KmipError(format!(
+                "Algorithm {:?} not supported while creating key",
+                &algorithm
+            )));
+        }
+
+        // Give keys a Krill specific but random name initially. Once we have created them we can determine the SHA-1
+        // of their X.509 SubjectPublicKeyInfo aka the Krill KeyIdentifier and use that in the name instead of the
+        // random component.
+
+        // The name given to a key is purely for our own use, the KMIP server doesn't care about it. We give keys a
+        // name that clearly indicates they relate to Krill as this may be helpful to the KMIP server operator. Once
+        // the key is created we rename it to include its Krill KeyIdentifier (aka the SHA-1 of the X.509
+        // SubjectPublicKeyInfo) so that we can relate the key back to its usage in Krill. We include the Unix seconds
+        // since 1970-01-01 timestamp in the name initially just as some rough at-a-glance indication of when it was
+        // created and to differentiate it from other keys with the same name (of which there should be none as they
+        // key should either be renamed after creation or should have been deleted at some point).
+        let prefix = format!("krill_new_key_{}", Timestamp::now());
+        let private_key_name = format!("{}_priv", prefix);
+        let public_key_name = format!("{}_pub", prefix);
+
+        // Create the RSA key pair
+        let kmip_key_pair_ids = self.create_rsa_key_pair(private_key_name, public_key_name)?;
+
+        // Prepare the new keys for use, and attempt to destroy them if anything goes wrong
+        let public_key = self
+            .prepare_keypair_for_use(&kmip_key_pair_ids.private_key_id, &kmip_key_pair_ids.public_key_id)
+            .or_else(|err| {
+                let _ = self.destroy_key_pair(&kmip_key_pair_ids, KeyStatus::Inactive);
+                Err(SignerError::KmipError(err.to_string()))
+            })?;
+
+        Ok((public_key, kmip_key_pair_ids))
+    }
+
+    /// Create an RSA key pair in the KMIP server.
+    fn create_rsa_key_pair(
+        &self,
+        private_key_name: String,
+        public_key_name: String,
+    ) -> Result<KmipKeyPairIds, SignerError> {
+        let (private_key_id, public_key_id) = self
+            .with_conn("create key pair", |conn| {
+                conn.create_rsa_key_pair(2048, private_key_name.clone(), public_key_name.clone())
+            })
+            .map_err(|err| SignerError::KmipError(format!("Failed to create RSA key pair: {}", err)))?;
+
+        let kmip_key_ids = KmipKeyPairIds {
+            public_key_id,
+            private_key_id,
+        };
+
+        Ok(kmip_key_ids)
+    }
+
+    /// Make the given KMIP private and public key pair ready for use by Krill.
+    fn prepare_keypair_for_use(&self, private_key_id: &str, public_key_id: &str) -> Result<PublicKey, SignerError> {
```

Comment:
why do we rename things using the key_id? I mean.. ultimately we have another thing mapping key_id to the actual keys and what not, so would it not be easier (more efficient in particular) to just have something unique like a uuid?

Reply:
ok, we can look at this later - and see if it impacts performance as well

Reply:
Speaking of performance we need to cache the results of key lookup.